### ISSUE TITLE
setWHashedPertubation without effect for chashed server policy

### ISSUE THREAD

**Issue by paddg** [PowerDNS/pdns#7186]
- Program: dnsdist
 - Issue type: Bug report

### Short description
dnsdist sends identical DNS queries to different backend servers after restart dnsdist. Or in case of multiple dnsdist instances each dnsdist has different hash results for equal queries. 

### Environment

 - Operating system: SLES12-SP3
 - Software version: dnsdist 1.3.3
 - Software source: self compiled

### Steps to reproduce
Configuration:
```
setWHashedPertubation(42)
setServerPolicy(chashed)
```

### Expected behaviour
I'd expect the equal queries to always be sent to the same back-end. Even if dnsdist was restartet. In the same way as with **_whashed_** policy.

### Actual behaviour
After a restart of dnsdist the queries will be forwarded to another server than before.

**Comment by rgacogne** [comment]
At a quick glance it should work since we are using the same hash perturbation (`g_hashperturb`) and it seems that the hashes are computed after the configuration has been parsed so `setWHashedPertubation()` should have already been applied.
@chbruyand Would you mind looking into this, please?

**Comment by chbruyand** [comment]
Backend's hashes distribution changes over restarts because their ID is generated randomly. You can instanciate backends with a pre-set UUID with something like:
```
newServer({                                                                                                                                                                                                                                              
      address="9.9.9.9",                                                                                                                                                                                                                                 
      id="90220ef3-4476-4d84-a484-80031d3b4dcd"                                                                                                                                                                                                          
})
```

It looks like the documentation is lacking infos about the `id` parameter of the `newServer()` function. Maybe we should also display the UUID in `showServers()` ?

**Comment by paddg** [comment]
Btw, that happened with our backends after
```
- setServerPolicy(whashed)
+ setServerPolicy(chashed)
```

![chashed](https://user-images.githubusercontent.com/1396936/48409483-3c44dc80-e73c-11e8-8d05-5175acb2a744.png)

**Comment by rgacogne** [comment]
> But why does it work with `whashed` then?

`whashed` does not use the `uuid`, `chashed` does.

**Comment by rgacogne** [comment]
I'm not certain I understand your graph, it seems like the CPU usage and latency increased for a while then got back to normal? If so that's what I would have expected since switching to `chashed` from `whashed` involves a different distribution of names so the recursor has a sudden increase of cache misses, but it should only happen once when switching from one policy to another, and then adding and removing servers should not trigger a redistribution.

**Comment by paddg** [comment]
The graph shows the change and the roll back. We have multiple dnsdists (x) and multiple backends (y) if each dnsdist has its own hash distribution, our overall backend cache has a y times redundancy.  And thus less efficiency. That means cpu and latency rises.

**Comment by rgacogne** [comment]
Ah right, thanks!

**Comment by pavel-odintsov** [comment]
Quite interesting issue, thanks for reporting it!

**Comment by chbruyand** [comment]
OK, so I don't think this is really an issue there. Nonetheless, the documentation lacks some clarifications that should be addressed in #7191 

Your feedback is welcome, and thanks again for your report!

**Comment by paddg** [comment]
Yes, I can confirm, it works with the proper "id" setting in newServer()

Thank you all!

**Comment by rgacogne** [comment]
Closing this since the missing documentation will be fixed in #7191. Thanks!
